code/modules/mob/living/silicon/ai/ai_say.dm

```diff
@@ -49,108 +49,137 @@
+	req.begin_async()
+	UNTIL(req.is_complete())
+	var/datum/http_response/res = req.into_response()
+	if(res.status_code == 200)
+		var/full_name_file = "data/vox_[vox_voice_number].wav"
+		// Process the audio to match SS13 AI effects.
+		shell("./data/ffmpeg.exe \
+			    -nostats -loglevel 0 \
+			    -i ./[full_name_file] \
+			    -i ./sound/effects/SynthImpulse.wav \
+			    -i ./sound/effects/RoomImpulse.wav \
+			    -filter_complex \" \
+			        \[0\] apad=pad_dur=2 \[in_1\]; \
+			        \[in_1\] asplit=2 \[in_1_1\] \[in_1_2\]; \
+			        \[in_1_1\] \[1\] afir=dry=10:wet=10 \[reverb_1\]; \
+			        \[in_1_2\] \[reverb_1\] amix=inputs=2:weights=8 1 \[mix_1\]; \
+			        \[mix_1\] asplit=2 \[mix_1_1\] \[mix_1_2\]; \
+			        \[mix_1_1\] \[2\] afir=dry=10:wet=10 \[reverb_2\]; \
+			        \[mix_1_2\] \[reverb_2\] amix=inputs=2:weights=10 1 \[mix_2\]; \
+			        \[mix_2\] equalizer=f=7710:t=q:w=0.6:g=-6,equalizer=f=33:t=q:w=0.44:g=-10 \[out\]; \
+        			\[out\] alimiter=level_in=1:level_out=1:limit=0.5:attack=5:release=20:level=disabled\" \
+    			-vn -y ./data/vox_[vox_voice_number].mp3")
+
+		if (!istype(SSassets.transport, /datum/asset_transport/webroot))
+			log_game("CDN not set up, VOX aborted.")
+			message_admins("CDN not set up, VOX aborted.")
+			fdel("data/vox_[vox_voice_number].wav")
+			fdel("data/vox_[vox_voice_number].mp3")
+			return
+		var/datum/asset_transport/webroot/WR = SSassets.transport
+		var/shit_fuck_ass = "data/vox_[vox_voice_number].mp3"
+		var/datum/asset_cache_item/ACI = new("[md5filepath(shit_fuck_ass)].mp3", file("data/vox_[vox_voice_number].mp3"))
+		ACI.namespace = "15aivox"
+		WR.save_asset_to_webroot(ACI)
+		var/url = WR.get_asset_url(null, ACI)
+
+		log_game("[key_name(speaker)] finished making a 15.AI announcement with the following message: [message]")
+		message_admins("[key_name(speaker)] finished making a 15.AI announcement with the following message: [message]")
+		speaker.say(";[message]")
+		// If there is no single listener, broadcast to everyone in the same z level
 		if(!only_listener)
 			// Play voice for all mobs in the z level
 			for(var/mob/M in GLOB.player_list)
-				if(M.can_hear() && (M.client.prefs.toggles & SOUND_ANNOUNCEMENTS))
-					var/turf/T = get_turf(M)
-					if(T.z == z_level)
-						SEND_SOUND(M, voice)
+				var/turf/T = get_turf(M)
+				if(T.z == z_level && M.can_hear())
+					M <<browse({"<META http-equiv="X-UA-Compatible" content="IE=edge"><audio autoplay><source src=\"[url]\" type=\"audio/mpeg\"></audio>"}, "window=vox_player&file=vox_player.htm")
+		else
+			only_listener <<browse({"<META http-equiv="X-UA-Compatible" content="IE=edge"><audio autoplay><source src=\"[url]\" type=\"audio/mpeg\"></audio>"}, "window=vox_player&file=vox_player.htm")
```

Comment:
I can guess that using `playsound` has the RAM implications of caching it forever. Can you add a comment explaining that please.

Reply:
We don't use playsound because that is laggy as fuck because it gets sent to each individual user one at a time. This is the fastest way to do it.

Reply:
playsound blocks network messages to the client while sending its chunks, and its chunking system uses small chunks and waits for the send to be confirmed by the client at the byond protocol level, and can cause the client to move to compressed network mode in byond which increases server load.

playsound also waits for the entire audio to download rather then stream it as it downloads

playsound also doesn't support mp3s

This uses the cdn because the cdn is bae